### DISCUSSION TITLE
Are contraints on function parameters possible? And how do I call the interfaces/functions of type classes?

### DISCUSSION THREAD

**Discussion by philschonholzer** [gcanti/fp-ts#1629] [Q&A] [answered]
Is it possible to write the following Haskell code with fp-ts? I want my function to accept any arguments that implement the Monoid type class.

```haskell
merge :: Monoid m => m -> m -> m
merge m1 m2 = m1 <> m2
```

How would I would this be written in TS?

```ts
const merge = (m1: ???, m2: ???): ??? => {
  return ???.concat(m1, m2);
}
```

The constraint is one thing, but how do I get my `m1` and `m2` to be a Monoid and how do I call `concat`?

Is this even possible? And how?

**Comment by xuanduc987** [comment, marked as answer]
Yes, it is possible but you will have to pass typeclass instance around.

```typescript
import { Monoid } from 'fp-ts/Monoid'
import { string } from 'fp-ts'

const merge = <A>(M: Monoid<A>) => (m1: A, m2: A): A => {
  return M.concat(m1, m2)
}

console.log(merge(string.Monoid)("abc", "xyz")) // => abcxyz
```

**Reply by philschonholzer** [reply]
Thanks! That makes sense.
